src/classes/CW_CampaignTriggerHandler.cls

```diff
@@ -0,0 +1,308 @@
+			Please note these master records are not persisted back, so that client gets a chance 
+			to post process them after rollup
+		*/ 
+		Sobject[] masters = LREngine.rollUp(ctx, objects);  
+		
+		// Persist the changes in master
+		if (!masters.isEmpty()) update masters;
+    }
+    
+    // Update Roadmap points where campaign previously had a roadmap; remove its campaigns points from the Roadmap (parent)
+    // Key of map is Roadmap Id, Value is list of Campaigns that previously had lookup to Roadmap key
+    public void updateRoadmapWhenCampaignsHaveBeenReparented(Map<Id,List<CW_Campaign__c>> reparentedObjectMap) {
+    	
+    	Map<Id,CW_Roadmap__c> roadmaps = new Map<Id,CW_Roadmap__c>([select Id, Points_Used__c from CW_Roadmap__c where Id in :reparentedObjectMap.keySet()]);
+		List<CW_Roadmap__c> roadmapsToUpdate = new List<CW_Roadmap__c>();
+		
+		for (Id roadmapId :reparentedObjectMap.keySet()) {
+			Double pointsToRemove = 0;
+			List<CW_Campaign__c> camps = reparentedObjectMap.get(roadmapId);
+			for (CW_Campaign__c c :camps) {
+				pointsToRemove += c.Points__c;
+			}
+			CW_Roadmap__c roadmap = roadmaps.get(roadmapId);
+			if (pointsToRemove!=0) {
+				roadmap.Points_Used__c -= pointsToRemove;
+				roadmapsToUpdate.add(roadmap);
+			}
+			
+		}
+		
+		if (!roadmapsToUpdate.isEmpty()) update roadmapsToUpdate;
+    }
+    
+    // update campaign stage on schedules with associated stage on campaign
+    public static void updateChildSchedulesWithCampaignStage(Map<Id,CW_Campaign__c> campaigns) {
+    	
+    	List<CW_Schedule__c> schedules = [select Id, Marketing_Campaign__c, Campaign_Stage__c from CW_Schedule__c where Marketing_Campaign__c in :campaigns.keySet()];
+    	
+    	// key: campaign Id, value: associated schedules
+    	Map<Id,List<CW_Schedule__c>> campaignScheduleMap = new Map<Id,List<CW_Schedule__c>>(); 
+    	for (CW_Schedule__c s :schedules) {
+    		List<CW_Schedule__c> campaignSchedules = campaignScheduleMap.get(s.Marketing_Campaign__c);
+    		if (campaignSchedules==null) campaignSchedules = new List<CW_Schedule__c>();
+    		campaignSchedules.add(s);
+    		campaignScheduleMap.put(s.Marketing_Campaign__c,campaignSchedules);
+    	}
+    	
+    	// update schedules with Campaign stage value
+    	List<CW_Schedule__c> schedulesToUpdate = new List<CW_Schedule__c>();
+    	for (CW_Campaign__c c :campaigns.values()) {
+    		List<CW_Schedule__c> campaignSchedules = campaignScheduleMap.get(c.Id);
+    		
+    		if (campaignSchedules!=null) {
+	    		for (CW_Schedule__c s :campaignSchedules) {
+	    			s.Campaign_Stage__c = c.Stage__c;
+	    			schedulesToUpdate.add(s);
+	    		}
+    		}
+    	}
+    	
+    	if (!schedulesToUpdate.isEmpty()) update schedulesToUpdate;
+    }
+	
+	public void stopNewCampaignCreationWhenRoadmapLocked(List<CW_Campaign__c> newObjects)
+	{
+		List<CW_Roadmap__c> roadmapDetails = [SELECT Id, Approval_Status__c
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.